src/colorizer/AnnotationData.ts

```diff
@@ -0,0 +1,222 @@
+   */
+  getLabelsAppliedToId(id: number): number[];
+
+  /**
+   * Returns all object IDs that the label has been applied to.
+   * @param labelIdx The index of the label to look up object IDs for.
+   * @returns an array of object IDs. Empty if the label has not been applied to
+   * any object IDs.
+   */
+  getLabeledIds(labelIdx: number): number[];
+
+  /**
+   * Returns a map from a frame number to the labeled object IDs present at that
+   * frame, represented as a record mapping label indices to IDs.
+   *
+   * The record's keys are an index of a label (as returned by `getLabels()`),
+   * and the values are arrays of object IDs present for that time frame that
+   * have that label applied.
+   *
+   * @param dataset The dataset to use for time information.
+   * @returns a map from time to a record of label indices to IDs.
+   * @example
+   * ```typescript
+   * // Let's say we have two labels (0 and 1). There are objects with IDs 1, 2,
+   * // and 3 at time 0, and IDs 4 and 5 at time 1.
+   * // Label 0 has been applied to objects 1, 2, 3, and 4.
+   * // Label 1 has been applied to objects 3 and 5.
+   *
+   * const timeToLabelMap = getTimeToLabelIdMap(some_dataset);
+   * timeToLabelMap.get(0); // { 0: [1, 2, 3], 1: [3] }
+   * timeToLabelMap.get(1); // { 0: [4], 1: [5]}
+   * ```
+   * */
+  getTimeToLabelIdMap(dataset: Dataset): Map<number, Record<number, number[]>>;
+
+  setLabelName(labelIdx: number, name: string): void;
+  setLabelColor(labelIdx: number, color: Color): void;
+  deleteLabel(labelIdx: number): void;
+
+  applyLabelToId(labelIdx: number, id: number): void;
+  removeLabelFromId(labelIdx: number, id: number): void;
+  toggleLabelOnId(labelIdx: number, id: number): void;
+}
+
+export class AnnotationData implements IAnnotationData {
+  private labelData: LabelData[];
+  private numLabelsCreated: number;
+  /**
+   * Cached mapping from time to label indices to IDs. Must be invalidated when
+   * labels are removed, or if annotations are applied to or removed from
+   * objects.
+   */
+  private timeToLabelIdMap: Map<number, Record<number, number[]>> | null;
+
+  constructor() {
+    this.labelData = [];
+    this.numLabelsCreated = 0;
+    this.timeToLabelIdMap = null;
+
+    this.getLabelsAppliedToId = this.getLabelsAppliedToId.bind(this);
+    this.getLabeledIds = this.getLabeledIds.bind(this);
+    this.getTimeToLabelIdMap = this.getTimeToLabelIdMap.bind(this);
+    this.getLabels = this.getLabels.bind(this);
+    this.createNewLabel = this.createNewLabel.bind(this);
+    this.setLabelName = this.setLabelName.bind(this);
+    this.setLabelColor = this.setLabelColor.bind(this);
+    this.deleteLabel = this.deleteLabel.bind(this);
+    this.toggleLabelOnId = this.toggleLabelOnId.bind(this);
+    this.applyLabelToId = this.applyLabelToId.bind(this);
+    this.removeLabelFromId = this.removeLabelFromId.bind(this);
+  }
+
+  // Getters
+
+  getLabels(): LabelData[] {
+    return [...this.labelData];
+  }
```

Comment:
1. Why does this copy, rather than just returning `this.labelData`?
2. Is it a problem that this shallow-copies? Modifying a member of the returned array will (if I'm not mistaken) modify the same `LabelData` object stored in `this.labelData`.

Reply:
I guess I wanted to protect the label data and make it readonly, but you're 100% right that you can still mutate the underlying objects. At the very least this prevents changing of the label ordering which would mess up all of the annotation data.